src/ripple/app/misc/impl/Manifest.cpp

```diff
@@ -127,184 +127,94 @@ Blob Manifest::getSignature () const
+    std::lock_guard<std::mutex> lock{read_mutex_};
+    auto const iter = map_.find (pk);
 
-    auto& value = map_[pk];
+    if (iter != map_.end ())
+        return iter->second.revoked ();
 
-    if (value.m)
-    {
-        Throw<std::runtime_error> (
-            "New trusted validator key already has a manifest");
-    }
-
-    value.comment = std::move(comment);
-}
-
-std::size_t
-ManifestCache::size () const
-{
-    std::lock_guard <std::mutex> lock (mutex_);
-    return map_.size ();
+    return false;
 }
 
 ManifestDisposition
-ManifestCache::canApply (PublicKey const& pk, std::uint32_t seq,
-    beast::Journal journal) const
+ManifestCache::canApply (
+    PublicKey const& pk,
+    std::uint32_t seq,
+    ValidatorList const& unl) const
 {
-    auto const iter = map_.find(pk);
-
-    if (iter == map_.end())
+    if (! unl.listed (pk) && ! unl.trustedPublisher (pk))
     {
         /*
             A manifest was received whose master key we don't trust.
             Since rippled always sends all of its current manifests,
             this will happen normally any time a peer connects.
         */
-        if (auto stream = journal.debug())
+        if (auto stream = j_.debug())
             logMftAct(stream, "Untrusted", pk, seq);
         return ManifestDisposition::untrusted;
     }
 
-    auto& old = iter->second.m;
+    auto const iter = map_.find(pk);
 
-    if (old  &&  seq <= old->sequence)
+    if (iter != map_.end() &&
+        seq <= iter->second.sequence)
     {
         /*
             A manifest was received for a validator we're tracking, but
-            its sequence number is no higher than the one already stored.
+            its sequence number is not higher than the one already stored.
             This will happen normally when a peer without the latest gossip
             connects.
         */
-        if (auto stream = journal.debug())
-            logMftAct(stream, "Stale", pk, seq, old->sequence);
+        if (auto stream = j_.debug())
+            logMftAct(stream, "Stale", pk, seq, iter->second.sequence);
         return ManifestDisposition::stale;  // not a newer manifest, ignore
     }
 
     return ManifestDisposition::accepted;
 }
 
-
 ManifestDisposition
 ManifestCache::applyManifest (
-    Manifest m, ValidatorList& unl, beast::Journal journal)
+    Manifest m, ValidatorList& unl)
 {
+    std::lock_guard<std::mutex> applyLock{apply_mutex_};
+
     /*
-        Move master public key from permanent trusted key list
-        to manifest cache.
+        before we spend time checking the signature, make sure we trust the
+        master key and the sequence number is newer than any we have.
     */
-    if (auto unlComment = unl.member (m.masterKey))
-    {
-        addTrustedKey (m.masterKey, *unlComment);
-        unl.removePermanentKey (m.masterKey);
-    }
+    auto const chk = canApply(m.masterKey, m.sequence, unl);
 
+    if (chk != ManifestDisposition::accepted)
```

Comment:
Remove the curly brackets. More generally, the `canApply` function should probably be:

    ManifestDisposition applicable (Manifest const& m, ValidatorList const& unl);